Спринт 13

polyakovanan opened the pull request:
Реализация восьми функциональностей в соответствии с задачами:
- Популярные фильмы
- Поиск
- Общие фильмы
- Отзывы
- Лента событий
- Рекомендации
- Удаление фильмов и пользователей
- Фильмы по режиссерам

polyakovanan committed:
Спринт 13.
Добавил функциональность по поиску общих лайкнутых фильмов

polyakovanan committed:
Спринт 13.
Добавил проверку на существование пользователей при запросе общих фильмов.
Поправил запрос, выдававший некорректный результат на предзаполненной БД

Viktor3642 committed:
Спринт 13. Итерация 1.
Добавил Функциональность «Отзывы».
1) Создал ветку add-reviews
2) Реализовал функции оценки полезности отзывов:
Для новых отзывов она начинается с 0
Для лайков увеличивается на 1
Для дизлайков уменьшается на 1
Отзывы сортируются по степени полезности
3) Реализовал функции получения отзывов по идентификатору фильма, если фильм не указан то все. Если кол-во не указано то 10.
4) Расширил схему SQL
5) Добавил тесты на поведение в оперативной памяти, и в базе данных.

Реализация обеспечивает функциональность для управления отзывами, включая их создание, обновление, удаление и оценку.
Постман и прошлые тесты проходит.

Viktor3642 committed:
Спринт 13. Итерация 1.
Добавил Функциональность «Отзывы».
1) Создал ветку add-reviews
2) Реализовал функции оценки полезности отзывов:
Для новых отзывов она начинается с 0
Для лайков увеличивается на 1
Для дизлайков уменьшается на 1
Отзывы сортируются по степени полезности
3) Реализовал функции получения отзывов по идентификатору фильма, если фильм не указан то все. Если кол-во не указано то 10.
4) Расширил схему SQL
5) Добавил тесты на поведение в оперативной памяти, и в базе данных.

Реализация обеспечивает функциональность для управления отзывами, включая их создание, обновление, удаление и оценку.
Постман и прошлые тесты проходит.

polyakovanan committed:
Merge pull request #5 from polyakovanan/add-common-films

Спринт 13. Функциональность «Общие фильмы»

polyakovanan committed:
Merge branch 'add-reviews' into add-feed

polyakovanan committed:
Спринт 13. Функциональность «Лента событий»
Добавил первое приближение ленты событий.
 Создал соответствующую модель, хранилища и репозитории. Добавил создание записей в связанных сервисах.
 Требуется перепроверить после фактического принятия в ветку develop задачи по «Отзывам»

Viktor3642 committed:
Merge remote-tracking branch 'origin/add-reviews' into add-reviews

Viktor3642 committed:
Спринт 13. Дорабатываю Функциональность «Отзывы».
1) Добавил юнит-тесты на удаление лайка/дизлайка в ReviewControllerTest.
2) В ReviewRepository в INSERT_QUERY убрал useful и 0.
3) В ReviewRepository объединил ADD_LIKE_QUERY и ADD_DISLIKE_QUERY в ADD_REACTION_QUERY.
4) В ReviewRepository объединил addLike с addDislike в addReaction и removeLike с removeDislike в removeReaction, обновил зависимости в DbReviewStorage.
5) В DbReviewStorage в update заменил оператор на лямба-выражение.
6) В InMemoryReviewStorage использовал .toList() вместо collect().

Viktor3642 committed:
Спринт 13. Дорабатываю Функциональность «Отзывы».
1) Добавил юнит-тесты на удаление лайка/дизлайка в ReviewControllerTest.
2) В ReviewRepository в INSERT_QUERY убрал useful и 0.
3) В ReviewRepository объединил ADD_LIKE_QUERY и ADD_DISLIKE_QUERY в ADD_REACTION_QUERY.
4) В ReviewRepository объединил addLike с addDislike в addReaction и removeLike с removeDislike в removeReaction, обновил зависимости в DbReviewStorage.
5) В DbReviewStorage в update заменил оператор на лямба-выражение.
6) В InMemoryReviewStorage использовал .toList() вместо collect().
7) В InMemoryReviewStorage убрал более неиспользуемый импорт.

ViktorSalk committed:
Merge pull request #7 from polyakovanan/add-reviews

Спринт 13. Функциональность " Отзывы" Итерация 1.

polyakovanan committed:
Merge branch 'develop' into add-feed

polyakovanan committed:
Спринт 13. Функциональность «Лента событий»
Изменил сортировку ленты с убывания по дате на возрастание. Так хотят тесты

polyakovanan committed:
Спринт 13. Функциональность «Лента событий»
фикс для чек-стайла

Viktor3642 committed:
Спринт 13. Функциональность «Рекомендации».

1) Создал ветку add-recommendations.
2) Реализовал функции простой рекомендательной системы для фильмов.
3) Реализовал в UserController конечную точку API GET /users/{id}/recommendations.
4) Добавил тесты.

Реализация обеспечивает функциональность для управления рекомендациями, находит пользователей со схожими вкусами, использует простой алгоритм сопоставления, возвращает рекомендации от наиболее похожего пользователя. Постман и прошлые тесты проходит.

polyakovanan committed:
Merge pull request #8 from polyakovanan/add-feed

Спринт 13. Функциональность «Лента событий»

Viktor3642 committed:
Спринт 13. Доработал Функциональность «Рекомендации».

1) Обновил InMemoryFilmControllerTest, InMemoryUserControllerTest, InMemoryReviewControllerTest добавив в аннотации SpringBootTest рекомендации.
В IDEA все тесты проходят.

Viktor3642 committed:
Merge remote-tracking branch 'origin/develop' into add-recommendations

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/film/InMemoryFilmControllerTest.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/review/InMemoryReviewControllerTest.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/user/InMemoryUserControllerTest.java

Viktor3642 committed:
Спринт 13. Доработал Функциональность «Рекомендации».

1) Обновил InMemoryFilmControllerTest, InMemoryUserControllerTest, InMemoryReviewControllerTest добавив в аннотации SpringBootTest рекомендации.
2) Объеденил изменения с веткой лента событий.
В IDEA все тесты проходят.

polyakovanan committed:
Спринт 13. Функциональность «Фильмы по режиссёрам»
Реализовал модель, контроллер, сервис, хранилища и репозиторий для сущности режиссера. Добавил таблицы в БД
Обновил сервис фильмов для работы с сущностью режиссеров
Добавил эндпоинт для поиска фильмов по режиссеру с сортировкой

umid1006 committed:
удаления фильма и пользователя по идентификатору

umid1006 committed:
удаления фильма и пользователя по идентификатору

umid1006 committed:
Merge remote-tracking branch 'origin/add-remove-endpoint' into add-remove-endpoint

Viktor3642 committed:
Спринт 13. Доработал Функциональность «Рекомендации».

1) Обновил в RecommendationService, collect() на toList,
Код стал более лаконичным, а список более защищенным от изменений.

ViktorSalk committed:
Merge pull request #9 from polyakovanan/add-recommendations

Спринт 13. Функциональность «Рекомендации».

umid1006 committed:
Merge remote-tracking branch 'origin/develop' into add-remove-endpoint

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java

umid1006 committed:
Merge remote-tracking branch 'origin/develop' into add-remove-endpoint

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/service/UserService.java

umid1006 committed:
Merge remote-tracking branch 'origin/add-remove-endpoint' into add-remove-endpoint

umid1006 committed:
Merge remote-tracking branch 'origin/add-remove-endpoint' into add-remove-endpoint

umid1006 committed:
Merge remote-tracking branch 'origin/add-remove-endpoint' into add-remove-endpoint

umid1006 committed:
Merge remote-tracking branch 'origin/develop' into add-remove-endpoint

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java

umid1006 committed:
Merge remote-tracking branch 'origin/develop' into add-remove-endpoint

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java

umid1006 committed:
Merge remote-tracking branch 'origin/develop' into add-remove-endpoint

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java

polyakovanan committed:
Merge branch 'develop' into add-director

polyakovanan committed:
Спринт 13. Функциональность «Фильмы по режиссёрам»
Добавил тесты для функциональности режиссеров

polyakovanan committed:
Спринт 13. Функциональность «Фильмы по режиссёрам»
Правки для чек-стайла

umid1006 committed:
Merge remote-tracking branch 'origin/develop' into add-remove-endpoint

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/UserController.java

polyakovanan committed:
Спринт 13. Функциональность «Фильмы по режиссёрам».
Исправления по замечаниям

umid1006 committed:
Merge pull request #10 from polyakovanan/add-remove-endpoint

удаления фильма и пользователя по идентификатору

polyakovanan committed:
Merge branch 'develop' into add-director

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/dal/repository/FilmRepository.java

polyakovanan committed:
Merge pull request #12 from polyakovanan/add-director

Спринт 13. Функциональность «Фильмы по режиссёрам»

umid1006 committed:
Добавение популярный фильмов

umid1006 committed:
Добавение популярных фильмов

umid1006 committed:
Добавение популярных фильмов

umid1006 committed:
Добавение популярных фильмов

umid1006 committed:
Добавение популярных фильмов

polyakovanan committed:
Спринт 13. Функциональность «Поиск».
Добавил эндпоинт для поиска фильмов по названию или имени режиссера со всеми соответствующими методами в контроллере, сервисе, хранилищах и репозитории.
Сделал так, чтобы эндпоинты могли принимать enum'ы
Добавил недостающие юнит-тесты

polyakovanan committed:
Merge pull request #15 from polyakovanan/add-search

Спринт 13. Функциональность «Поиск».

polyakovanan committed:
Merge branch 'develop' into add-most-populars

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/storage/dal/repository/FilmRepository.java
#	src/test/java/ru/yandex/practicum/filmorate/controller/film/FilmControllerTest.java

polyakovanan committed:
Спринт 13. Функциональность «Популярные фильмы».
Переиспользовал наработки @umid1006, чтобы добавить фильтрацию по жанру и году выпуска при поиске популярных фильмов.
Сделал параметр count необязательным.

Провел тотальное тестирование проекта на этой ветке, так как эта задача последняя и внес правки в весь функционал, который не проходил проверки

polyakovanan committed:
Спринт 13. Функциональность «Популярные фильмы».
Обновил схему БД

polyakovanan committed:
Merge pull request #16 from polyakovanan/add-most-populars

Спринт 13. Функциональность «Популярные фильмы»

Umka7389 commented on src/main/java/ru/yandex/practicum/filmorate/model/event/Event.java:
```diff
@@ -0,0 +1,16 @@
+@Data
+@Builder
+public class Event {
+    long eventId;
```
Модификаторы доступа не нужны?

polyakovanan replied:
Добавили модификаторы доступа к полям всех классов моделей

Umka7389 commented on src/main/java/ru/yandex/practicum/filmorate/service/DirectorService.java:
```diff
@@ -0,0 +1,70 @@
+            directorStorage.delete(id);
+            return;
+        }
+        log.error(String.format(NOT_FOUND_MESSAGE, id));
```
Попробуйте посмотреть в сторону конструкции orElseThrow в Optional. Лог дополнительный вам не нужен у вас то же самое сообщение в эксепшене, таким образом можно будет упростить код. Применимо ко многим местам в вашем коде

polyakovanan replied:
Заменили .get() в ветвлениях на orElseThrow везде, где есть связанная выброска исключения.
Интересно, что Sonar ругается на использование orElseThrow без использования возвращаемого значения, но с Atlassian признал это багом и завел тикет, который не могут решить уже 2 года))
https://sonarsource.atlassian.net/browse/SONARJAVA-4431

Umka7389 commented on src/main/java/ru/yandex/practicum/filmorate/storage/film/InMemoryFilmStorage.java:
```diff
@@ -32,23 +32,80 @@ public Optional<Film> getById(long id) {
+                .filter(f -> commonFilms.contains(f.getId()))
+                .forEach(film -> likedFilms.put(film.getId(), 0L));
+
+        likedFilms.putAll(likeStorage.findAll().stream().filter(like -> commonFilms.contains(like.getFilmId()))
```
Заметил, что у вас много где используется подход, что вы из БД получается все данные, а потом уже фильтруете. Так делать не стоит, это явно ударит по производительности. Попробуйте переписать и использовать запросы к БД с условиями и получать уже готовые коллекции, где это допустимо

polyakovanan replied:
Исключили использование полных выборок везде, где позволяет логика кода и заменили на выборки с фильтрацией.
Неизменным осталась только функциональность рекомендаций, поскольку для своей работы она действительно обходит все имеющиеся лайки: сначала целевого пользователя для выборка его фильмов, а потом всех пользователей кроме целевого, для поиска фильмов похожих на ранее выбранные

Umka7389 reviewed (commented):
Молодцы, отличная работа! Есть небольшие рекомендации и замечания

polyakovanan committed:
Спринт 13. Итерация ревью 1.
Внесены изменения по замечаниям от заказчика.
1) Поля классов моделей данных теперь имеют модификатор доступа private
2) Проверка на существование записей в хранилищах теперь осуществляется через orElseThrow класс Optional
3) Выборки всех данных из хранилищ с последующих обходом теперь заменены выборками с фильтрацией

polyakovanan committed:
Merge pull request #18 from polyakovanan/review-iteration-1

Спринт 13. Итерация ревью 1.

Umka7389 reviewed (commented):
Молодцы, отличная работа!
